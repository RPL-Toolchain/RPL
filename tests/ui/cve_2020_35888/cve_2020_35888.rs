//@ ignore-on-host
use std::alloc::{alloc, Layout};

pub struct Array<T> {
    size: usize,
    ptr: *mut T,
}

impl<T> Array<T>
where
    T: Clone,
{
    pub fn new_from_template(size: usize, template: &T) -> Self {
        let objsize = std::mem::size_of::<T>();
        let layout = Layout::from_size_align(size * objsize, 8).unwrap();
        let ptr = unsafe { alloc(layout) as *mut T };
        for i in 0..size {
            unsafe {
                (*(ptr.wrapping_offset(i as isize))) = template.clone();
            }
        }
        Self { size, ptr }
    }
}
/* 
#[derive(Clone, Debug)]
struct DropDetector(u32);

impl Drop for DropDetector {
    fn drop(&mut self) {
        println!("Dropping value: {} at {:?}", self.0, self as *const _);
    }
}
fn main() {
    let array = Array::new_from_template(2, &DropDetector(12345));
    for i in 0..array.size {
        // drop elements
        unsafe {
            let ptr = array.ptr.wrapping_offset(i as isize);
            std::ptr::drop_in_place(ptr);
        }
    }
} */