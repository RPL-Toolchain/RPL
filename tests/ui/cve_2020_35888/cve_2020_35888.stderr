error: Possibly dropping an uninitialized value
  --> tests/ui/cve_2020_35888/cve_2020_35888.rs:18:17
   |
LL |                 (*(ptr.wrapping_offset(i as isize))) = template.clone();
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 1 previous error

